core/trino-main/src/main/java/io/trino/metadata/MetadataListing.java

```diff
@@ -132,22 +134,54 @@ private MetadataListing() {}
 
     public static Map<SchemaTableName, List<ColumnMetadata>> listTableColumns(Session session, Metadata metadata, AccessControl accessControl, QualifiedTablePrefix prefix)
     {
-        Map<SchemaTableName, List<ColumnMetadata>> tableColumns = metadata.listTableColumns(session, prefix).entrySet().stream()
-                .collect(toImmutableMap(entry -> entry.getKey().asSchemaTableName(), Entry::getValue));
+        Map<SchemaTableName, Optional<List<ColumnMetadata>>> tableColumns = metadata.listTableColumns(session, prefix).values().stream()
+                .flatMap(List::stream)
+                .collect(toImmutableMap(TableColumnsMetadata::getTable, TableColumnsMetadata::getColumns));
+
         Set<SchemaTableName> allowedTables = accessControl.filterTables(
                 session.toSecurityContext(),
                 prefix.getCatalogName(),
                 tableColumns.keySet());
 
         ImmutableMap.Builder<SchemaTableName, List<ColumnMetadata>> result = ImmutableMap.builder();
-        for (Entry<SchemaTableName, List<ColumnMetadata>> entry : tableColumns.entrySet()) {
+        for (Entry<SchemaTableName, Optional<List<ColumnMetadata>>> entry : tableColumns.entrySet()) {
             if (!allowedTables.contains(entry.getKey())) {
                 continue;
             }
-            List<ColumnMetadata> columns = entry.getValue();
+
+            QualifiedObjectName originalTableName = new QualifiedObjectName(prefix.getCatalogName(), entry.getKey().getSchemaName(), entry.getKey().getTableName());
+            Optional<List<ColumnMetadata>> columnsOptional = entry.getValue();
+            List<ColumnMetadata> columns;
+
+            if (columnsOptional.isPresent()) {
```

Comment:
Could `metadata.listTableColumns` handle column listing itself?

Reply:
we had considered that. From PR description: 

```
The list of table names comes solely from the source catalog, thus 
access control filterTables is only called for the source catalog. The 
target catalog is not involved at all with listing names. Because the 
list of columns comes from the target catalog, the access control 
filterColumns needs to be called for the target catalog for redirections. 
We do this by modifying MetadataListing.listTableColumns to do the 
redirection logic for listing columns, rather than doing it in MetadataManager.
```